File: pyinterpreter/interpreter/parser.py
from interpreter.ast import Number, BinOp, UnOp
from interpreter.lexer import Lexer
from interpreter.token import TokenType, Token


class Parser:
    def __init__(self):
        self._current_token = None
        self._lexer = Lexer()

    def check_token(self, type_: TokenType):
        if self._current_token.type_ == type_:
            self._current_token = self._lexer.next()
        else:
            raise SyntaxError("invalid token order")

    def factor(self):
        token = self._current_token
        if token.type_ == TokenType.NUMBER:
            self.check_token(TokenType.NUMBER)
            return Number(token)

        if token.type_ == TokenType.L_PAREN:
            self.check_token(TokenType.L_PAREN)
            result = self.expr()
            self.check_token(TokenType.R_PAREN)
            return result

        if token.type_ == TokenType.OPERATOR:
            self.check_token(TokenType.OPERATOR)
            return UnOp(self.factor(), token)
        raise SyntaxError("Invalid factor")

    def term(self):
        result = self.factor()
        while self._current_token and (self._current_token.type_ == TokenType.OPERATOR):
            if self._current_token.value not in ["*", "/"]:
                break
            token = self._current_token
            self.check_token(TokenType.OPERATOR)
            return BinOp(result, token, self.factor())
        return result

    def expr(self):
        result = self.term()
        while self._current_token and (self._current_token.type_ == TokenType.OPERATOR):
            token = self._current_token
            self.check_token(TokenType.OPERATOR)
            result = BinOp(result, token, self.term())
        return result

    def parse(self, code):
        self._lexer.init(code)
        self._current_token = self._lexer.next()
        return self.expr()

File: pascalInterpreter/src/main/kotlin/interpreter/Parser.kt
package interpreter

import Exceptions.SyntaxException

class Parser {
    private var currentToken: Token? = null
    private var lexer = Lexer()

    private fun check_token(type: TokenType) {
        if (currentToken?.type == type) {
            currentToken = lexer.next()
        } else {
            throw SyntaxException("Invalid token order")
        }
    }

    private fun factor(): Node {
        val token = currentToken
        when (token!!.type) {
            TokenType.NUMBER -> {
                check_token(TokenType.NUMBER)
                return PascalNumber(token)
            }

            TokenType.L_PAREN -> {
                check_token(TokenType.L_PAREN)
                val result = expr()
                check_token(TokenType.R_PAREN)
                return result
            }

            TokenType.OPERATOR -> {
                check_token(TokenType.OPERATOR)
                return UnOp(token, factor())
            }

            TokenType.ID -> {
                check_token(TokenType.ID)
                return Identifier(token)
            }

            else -> {
                throw SyntaxException("Invalid factor")
            }
        }
    }

    private fun variable(): Token {
        val result = currentToken
        check_token(TokenType.ID)
        return result!!
    }

    private fun assignment(): Node {
        val idToken = variable()
        check_token(TokenType.ASSIGN)
        return Variable(idToken, expr())
    }

    private fun statement_list(): Node {
        var result = statement()
        if (currentToken!!.type == TokenType.SEMI) {
            check_token(TokenType.SEMI)
            result = Semi(result, statement_list())
        }
        return result
    }

    private fun complex_statement(): Node {
        check_token(TokenType.BEGIN)
        val result = statement_list()
        check_token(TokenType.END)
        return result
    }

    private fun empty(): Node {
        return Empty()
    }

    private fun program(): Node {
        val result = complex_statement()
        check_token(TokenType.DOT)
        return result
    }

    private fun statement(): Node {
        return when (currentToken?.type) {
            TokenType.BEGIN -> complex_statement()
            TokenType.ID -> assignment()
            TokenType.END -> empty()
            else -> throw SyntaxException("Invalid token")
        }
    }

    private fun term(): Node {
        val result = factor()
        while (currentToken!!.type == TokenType.OPERATOR) {
            if (currentToken!!.value !in "*/") {
                break
            }
            val token = currentToken
            check_token(TokenType.OPERATOR)
            return BinOp(result, token!!, term())
        }
        return result
    }

    private fun expr(): Node {
        var result = term()
        while (currentToken!!.type == TokenType.OPERATOR) {
            val token = currentToken
            check_token(TokenType.OPERATOR)
            result = BinOp(result, token!!, term())
        }
        return result
    }

    fun parse(code: String): Node {
        lexer.init(code)
        currentToken = lexer.next()
        return program()
    }
}